### DISCUSSION TITLE
Content Security Policy Header

### DISCUSSION THREAD

**Discussion by aureateflux** [photoprism/photoprism#2714] [Q&A]
Does anyone have an example of a CSP header setting that will work with Photoprism?

I have been trying to improve the security of the services exposed on my domain by updating my security headers in Traefik.  The Content Security Policy is the last one standing, and I've discovered that setting it wrong *thoroughly* breaks Photoprism's web ui.

I've come up with some headers that mostly work. I'm a little uncertain about the security consequences of setting 'unsafe-inline' sources, but I don't think there's an alternative.  The header I came up with also blocks something that might be a tracking pixel that might be needed for...something? I couldn't figure out how to allow this specific resource without allowing anything from data: to go through.  

Anyway, my Traefik CSP header is as follows, feedback and recommendations absolutely requested:

contentSecurityPolicy: >-
  connect-src 'self';
  default-src [DOMAIN.com];
  media-src 'self';
  font-src 'self';
  manifest-src 'self';
  img-src 'self';
  script-src 'self' 'unsafe-inline';
  style-src 'self' 'unsafe-inline';

**Comment by lastzero** [comment]
AFK right now, so I can't send you our configuration, but you could have a look at the HTTP headers of our demo. It includes a few additional domains, which you don't need. Most importantly, you need to whitelist maptiler for the maps to work.

PS: Feel free to suggest improvements and/or send a pull request for our docs to help other users with the same question.

**Reply by aureateflux** [reply]
Thanks for the response!

I hadn't thought of looking at the HTTP headers on the demo! I had a look at it, and it was very helpful. I wanted a more restrictive default-src though (although do I even need one when I'm providing values for basically all the directives? haha), so after some trial and error I came up with this:

frame-ancestors 'none';
connect-src 'self' *.maptiler.com;
default-src [DOMAIN].com *.[DOMAIN].com;
media-src 'self' data:;
font-src 'self';
manifest-src 'self';
worker-src 'self' blob:;
img-src 'self' data:;
script-src 'self' 'unsafe-inline' 'unsafe-eval';
style-src 'self' 'unsafe-inline';
object-src 'none';

This worked everywhere I tested; initially I thought *.stripe.com might need to be whitelisted, but I couldn't find any resources that were rejected because of it, so I assume it's only used for payment processing on the non-demo parts of the photoprism.app site. I also couldn't find anything that actually requires 'unsafe-eval' so if that's not needed for the Photoprism deployment, it can be removed (putting it in script-src was mostly just guessing where it would most logically be needed).

"blob:" seems to only be needed for the map, so that needs to be whitelisted in worker-src along with whitelisting *.maptiler.com in connect-src. I thought it might be needed for uploads or downloads, but they seem to be working OK.

If anyone else uses these headers or streamlines them or has any notes on them, please share. It's entirely possible that something here still breaks something in Photoprism that I didn't think to test, so use with care.

I hesitate to suggest adding these specific headers to documentation until they've been vetted by people with more expertise than I have, but maybe posting some sample header configs would be helpful in the documentation. From what I could tell, these were the only class of security headers that created any real conflicts with Photoprism, but don't take that to the bank.

**Reply by lastzero** [reply]
Check if websockets still work with this. Not sure why we had to add unsafe eval, could have been needed for a different web app like weblate.

**Reply by aureateflux** [reply]
Websockets were connecting, but I'm not sure about the best way to actively test it. I did drop unsafe-eval and things worked fine.

**Reply by lastzero** [reply]
Open the browser console and check the logs and network tabs for errors and warnings.
